src/ImageSharp/Processing/Processors/Normalization/AutoLevelProcessor{TPixel}.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) Six Labors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Linq;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Normalization
+{
+    /// <summary>
+    /// Normalize an image by stretching the dynamic range to full contrast
+    /// Applicable to an <see cref="Image"/>.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal class AutoLevelProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        private readonly int luminanceLevels;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoLevelProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="configuration">The configuration which allows altering default behaviour or extending the library.</param>
+        /// <param name="source">The source <see cref="Image{TPixel}"/> for the current processor instance.</param>
+        /// <param name="sourceRectangle">The source area to process for the current processor instance.</param>
+        public AutoLevelProcessor(
+            Configuration configuration,
+            Image<TPixel> source,
+            Rectangle sourceRectangle)
+            : base(configuration, source, sourceRectangle)
+        {
+            // TODO I don't know how to get channel bit depth for non-grayscale types
+            if (!(typeof(TPixel) == typeof(L16) || typeof(TPixel) == typeof(L8)))
```

Comment:
I think it would be ok to take the `PixelType.BitsPerPixel` and divide it by `Unsafe.SizeOf<T>()`. That will give you a per-pixel-component average (_since all our pixel formats are blittable_) which should be close enough.

Reply:
If I would try that with `L16`

`Unsafe.SizeOf<L16>()` would return `2`
`BitsPerPixel` would return `16`

It's half of what it needs to be?

Or did you mean it's an operation only for non-luminescence types?

Reply:
Ah right yep, my bad We need a `ChannelCount` property on `PixelTypeInfo`. 

@antonfirsov This would be a breaking change (_which we can probably allow for 1.1 release_) but do you have any ideas how we could introduce it in a non breaking way?

Reply:
A hack you could do just now is something similar to what we do in the png encoder, I.E use a switch for known types and a default for unknown.
 
 https://github.com/SixLabors/ImageSharp/blob/a8cb71158493096121439d937b7e79aa5ea2727a/src/ImageSharp/Formats/Png/PngEncoderOptionsHelpers.cs#L159-L174